tests/hipify-clang/cuRAND/benchmark_curand_generate.cpp

```diff
@@ -0,0 +1,393 @@
+#include <vector>
+#include <string>
+#include <chrono>
+#include <numeric>
+#include <utility>
+#include <algorithm>
+
+#include "cmdparser.hpp"
+// CHECK: #include <hip/hip_runtime.h>
+#include <cuda_runtime.h>
+// CHECK: #include <hiprand.h>
+#include <curand.h>
+
+// CHECK: if((x)!=hipSuccess) {
+#define CUDA_CALL(x) do { if((x)!=cudaSuccess) { \
+    printf("Error at %s:%d\n",__FILE__,__LINE__);\
+    exit(EXIT_FAILURE);}} while(0)
+// CHECK: if((x)!=HIPRAND_STATUS_SUCCESS) {
+#define CURAND_CALL(x) do { if((x)!=CURAND_STATUS_SUCCESS) { \
+    printf("Error at %s:%d\n",__FILE__,__LINE__);\
+    exit(EXIT_FAILURE);}} while(0)
+
+#ifndef DEFAULT_RAND_N
+const size_t DEFAULT_RAND_N = 1024 * 1024 * 128;
+#endif
+
+// CHECK: typedef hiprandRngType_t rng_type_t;
+typedef curandRngType rng_type_t;
+
+// CHECK: using generate_func_type = std::function<hiprandStatus_t(hiprandGenerator_t, T *, size_t)>;
+template<typename T>
+using generate_func_type = std::function<curandStatus_t(curandGenerator_t, T *, size_t)>;
+
+template<typename T>
+void run_benchmark(const cli::Parser& parser,
+                   const rng_type_t rng_type,
+                   generate_func_type<T> generate_func)
+{
+    const size_t size = parser.get<size_t>("size");
+    const size_t trials = parser.get<size_t>("trials");
+
+    T * data;
+    // CHECK: CUDA_CALL(hipMalloc((void **)&data, size * sizeof(T)));
+    CUDA_CALL(cudaMalloc((void **)&data, size * sizeof(T)));
+
+    // CHECK: hiprandGenerator_t generator;
+    // CHECK: CURAND_CALL(hiprandCreateGenerator(&generator, rng_type));
+    curandGenerator_t generator;
+    CURAND_CALL(curandCreateGenerator(&generator, rng_type));
+
+    const size_t dimensions = parser.get<size_t>("dimensions");
+    // CHECK: hiprandStatus_t status = hiprandSetQuasiRandomGeneratorDimensions(generator, dimensions);
+    // CHECK: if (status != HIPRAND_STATUS_TYPE_ERROR)
+    curandStatus_t status = curandSetQuasiRandomGeneratorDimensions(generator, dimensions);
+    if (status != CURAND_STATUS_TYPE_ERROR) // If the RNG is not quasi-random
+    {
+        CURAND_CALL(status);
+    }
+
+    // Warm-up
+    for (size_t i = 0; i < 5; i++)
+    {
+        CURAND_CALL(generate_func(generator, data, size));
+    }
+    // CHECK: CUDA_CALL(hipDeviceSynchronize());
+    CUDA_CALL(cudaDeviceSynchronize());
+
+    // Measurement
+    auto start = std::chrono::high_resolution_clock::now();
+    for (size_t i = 0; i < trials; i++)
+    {
+        CURAND_CALL(generate_func(generator, data, size));
+    }
+    // CHECK: CUDA_CALL(hipDeviceSynchronize());
+    CUDA_CALL(cudaDeviceSynchronize());
+    auto end = std::chrono::high_resolution_clock::now();
+    std::chrono::duration<double, std::milli> elapsed = end - start;
+
+    std::cout << std::fixed << std::setprecision(3)
+              << "      "
+              << "Throughput = "
+              << std::setw(8) << (trials * size * sizeof(T)) /
+                    (elapsed.count() / 1e3 * (1 << 30))
```

Comment:
it seems `elapsed.count()` can possible return zero so it may be div/0 here and at line 110?

Reply:
You'd better file an issue to [rocRAND](https://github.com/ROCmSoftwarePlatform/rocRAND/issues), as those benchmarks are taken from there.

Reply:
It's highly unlikely that `elapsed.count()` will be zero (IIRC, on most platforms `std::chrono::high_resolution_clock` is steady), and even if that happens the expression will evaluate to `inf`. I don't think that's a problem, because if  `elapsed.count()` is zero, then the result is useless anyway and something must be wrong.

Reply:
Well, I guess this might be merged as is for now.